### ISSUE TITLE
Rust support

### ISSUE THREAD

**Issue by LogicAndTrick** [kaitai-io/kaitai_struct#22]
Adding this so it can be tracked, but I plan on doing most of this myself. I've not yet had a need for Rust but I've been wanting to try it out anyway, figured this is a good excuse.

I've created a runtime for Rust: https://github.com/kaitai-io/kaitai_struct_rust_runtime

I'll be adding matching tests and a compiler for Rust in the future.

Note that I don't know Rust at all so I'm learning as I go. If anybody has coded with it before I would very much appreciate advice, suggestions, or even a code review of the runtime!

**Comment by GreyCat** [comment]
Wow, that's cool :) May be you should post it to some Rust-related mailing lists / communities, to attract more peer review? I don't know a thing about Rust myself ;)

**Comment by LogicAndTrick** [comment]
After working with the language a bit, I've come to realise how difficult of a task this would be. Rust is certainly an interesting language to work with, but it comes with a huge list of new challenges that don't exist in other languages. I don't see myself being able to overcome them, so I'm going to quit.

There's a whole stack of architecture questions that are easy to answer in other languages, but actually very complex in Rust. The design of code generated by KS follows the easy patterns because that's how pretty much every other language works. But Rust is unique in many areas.

Rust has what they call a [foreign function interface](https://doc.rust-lang.org/book/ffi.html) which lets you interact with C libraries, perhaps that's a more reasonable approach for Rust. But for now, I'm going to close this because I'm not able to make any more progress.

(I enjoyed experimenting with Rust: it sort of feels like a modern approach to C. The memory safety syntax is very complex for beginners, but the rest of language is quite simple and expressive. The tools are a pleasure to use as well. Would definitely recommend anyone to spend a weekend with it if they've ever been curious about it.)

**Comment by GreyCat** [comment]
Maybe someone would return to Rust implementation in future, so I thought it might be interesting to document your findings or at least map out possible obstacles on the path. What does the language lack that impedes KS implementation? Something about the type system? Memory management? Strict compile-time checks? Lack of "everything is an expression"? Something else?

**Comment by LogicAndTrick** [comment]
The main pain point I discovered is around the references to `io`, `root`, and `parent` in each object. If they could be removed somehow, I think I could probably get everything else going. From what I've found online, root/parent/self references require some pretty complex type chains, and the generated code would probably be pretty crazy as well. I don't see myself ever becoming proficient enough in Rust to be able to get that going correctly, and even if you could do it, the result would probably not be something that anybody would want to actually use.

If instance methods required these three pointers to be passed as arguments, then it might be more possible. For example:

```rust
// hard to do
fn get_calculated_instance(self) { ... }

// probably easier, but consumer must pass root/parent manually
fn get_calculated_instance(self, io: Stream, parent: Struct, root: Struct) { ... }
```

From what it seems, Rust is best approached by creating an architecture that specifically fits Rust's "memory safety first" style. Trying to modify an existing architecture so that it works in Rust is very difficult.

**Comment by GreyCat** [comment]
That's kind of strange. I guess that Rust has something that resembles an object (i.e. bundle of data + methods) and I don't see any difference in storing some "useful" data, i.e. attributes / instances vs storing a reference (pointer?) to `io` object that should be used for structure parsing, or `parent`, or `root`.

But that said, I know almost nothing about Rust, so I can only make theories.

**Comment by GreyCat** [comment]
I'm reorganizing all "new language support" issues, adding relevant tag. I'll reopen this one, as a few people asked for Rust support again, and it's better to keep it open, so it would be easier to find and people won't create duplicates again and again.

**Comment by berkus** [comment]
Yes, please, very much need Rust support from Kaitai!

**Comment by GreyCat** [comment]
@berkus You probably know of http://doc.kaitai.io/new_language.html — let's just follow. I'd be happy to launch a Rust-targeting branch as soon as we'll get basic `hello_world.rs` and basic tests for it.

**Comment by berkus** [comment]
No, this is new to me - I just started actually using Kaitai today, loving it so far.

Will take a look!

**Comment by LogicAndTrick** [comment]
I'm going to start tentatively taking a second look at this one because the second edition of the Rust book has a [dedicated chapter on reference cycles](https://doc.rust-lang.org/book/second-edition/ch15-06-reference-cycles.html) which might help me solve the roadblocks I was hitting last year. I still have no experience with actual proper Rust code so if anybody has more experience with the language, I'd love some assistance!

Question for @GreyCat: I've manually assembled the code for a `hello_world.rs` file which passes basic tests, where would I put it? I don't think it's ready for CI scripts at this point, I still have to figure out the best way to separate the test project from the code project.

For the time being, it'd be good to get some feedback so here's what I've got so far for hello_world:

```rust
extern crate kaitai_struct;

use kaitai_struct::{Stream, Struct};

use std::io::{Cursor, Read, Seek};
use std::fs::File;

struct HelloWorld {
    one: u8,
}

impl Struct for HelloWorld {

    // Rust requires all fields to be initialised so we auto-generate an empty ctor
    fn empty() -> HelloWorld {
        HelloWorld { one: 0 }
    }

    // Passes any io::Error through to the caller using the ? operator
    fn from_file(path: &str) -> std::result::Result<HelloWorld, std::io::Error> {
        let mut buf = File::open(path)?;
        Ok(HelloWorld::new(
            &mut buf,
            &None::<Box<Struct>>,
            &None::<Box<Struct>>,
        ))
    }

    // Constructor from a stream
    // Box<> References to parent and root are temporary and the types will change.
    fn new<S: Stream>(
        stream: &mut S,
        parent: &Option<Box<Struct>>,
        root: &Option<Box<Struct>>,
    ) -> HelloWorld {
        let mut this = HelloWorld::empty();
        this.read(stream, &parent, &root);
        this
    }

    // Read takes a reference to stream/parent/root, this may change later.
    fn read<S: Stream>(
        &mut self,
        stream: &mut S,
        parent: &Option<Box<Struct>>,
        root: &Option<Box<Struct>>,
    ) {
        self.one = stream.read_u1().unwrap();
    }
}

#[cfg(test)]
mod tests {
    use kaitai_struct::Struct;
    use HelloWorld;

    #[test]
    fn test_hello_world() {
        let hw = HelloWorld::from_file("./src/fixed_struct.bin")
            .expect("File not found!");
        assert_eq!(hw.one, 0x50);
    }
}
```

I'm going to start working out the `parent` and `root` references, if I can figure that out then I'm hoping that it should be pretty basic stuff after that.

**Comment by GreyCat** [comment]
Looks really cool :)

As for code location, I'd just leave it here at the moment. It's not like putting it somewhere into the tests repo would help anything — we need test specs, we need the parser code to be generated, etc, so the hand-written generation result doesn't really help much in terms of CI or stuff like that.

May be we could ask some Rust online communities for some input / suggestions on this one?

**Comment by berkus** [comment]
I've implemented some parts in Rust, inspired by kaitai cpp generator [here](https://github.com/berkus/carma/blob/rust/src/support/resource.rs#L72).

Maybe this gives you new ideas (matching .ksy file is [here](https://github.com/berkus/carma/blob/rust/kaitai/carma.ksy))

**Comment by LogicAndTrick** [comment]
Neat, you wouldn't by any chance know a good way to represent the parent/root references in Rust? I have looked at it a few times and it just seems to create a painful chain of types like `Option<Weak<RefCell<Rc<T>>>>` and the interface it creates seems pretty unreasonable for somebody to want to use.

**Comment by berkus** [comment]
It's pretty painful because of borrow checker. Usually peeps do it via a Vec<T> and integer indices pointing to parent or children from a T.

Probably some rust tree package like ego_tree or idtree may give you help?

Otherwise I believe `RefCell<Rc<T>>` is the only way (but i'm no expert - been doing rust for barely two weeks now)

**Comment by CWood1** [comment]
Since I don't know the status of this, I decided to start an alternate implementation. Still very much a WIP, but hello world now successfully compiles to this:

`// This is a generated file! Please edit source .ksy file and use kaitai-struct-compiler to rebuild

use std::{
    option::Option,
    boxed::Box,
    io::Result
};

use kaitai_struct::{
    KaitaiStream,
    KaitaiStruct
};

pub struct HelloWorld {
    pub one: u8,
}

impl KaitaiStruct for HelloWorld {
    fn new<S: KaitaiStream>(stream: &mut S,
                            _parent: &Option<Box<KaitaiStruct>>,
                            _root: &Option<Box<KaitaiStruct>>)
                            -> Result<Self>
        where Self: Sized {
        let mut s = Self {
            one: 0,
        };

        s.read(stream, _parent, _root)?;

        Ok(s)
    }

    fn read<S: KaitaiStream>(&mut self,
                             stream: &mut S,
                             _parent: &Option<Box<KaitaiStruct>>,
                             _root: &Option<Box<KaitaiStruct>>)
                             -> Result<()>
        where Self: Sized {
        self.one = stream.read_u1()?;

        Ok(())
    }
}
`

I've tested this and confirmed it all working. My next step is to port some of the test suite over, writing test harnesses for Rust. From there, I can get them all passing, one by one.

**Comment by GreyCat** [comment]
@CWood1 Thanks, looks really cool! Note that:

* Nowadays there is a roadmap for implementation of new languages: looks [like this](https://github.com/kaitai-io/kaitai_struct/issues/146#issuecomment-326558115), and can be generated [using these scripts](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/doc).

* There is really no need to port majority of test specs manually: one can implement expression support for target language, and then harness the power of KST (see #253). The translator project is [here](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/translator) and once you'll add support for the new language there, porting many tests is a breeze, as you get it to generate specs automatically.

**Comment by CWood1** [comment]
Oh wow, that all looks much easier than I'd anticipated! Is there any documentation on generating and running the tests for a given language automatically? Tried to figure it out from travis.yml, but that does strange things with packages which won't work on my Arch installation.

Thanks for the help, I'll take a look at getting the test translator ported next.

**Comment by GreyCat** [comment]
Unfortunately, not now, as this whole thing was literally implemented a month or so ago. Travis yaml won't help much here, as specs are to autogenerated on the fly during normal build process.

To run test translator, you'd generally want to run:

1. Go to compiler dir
2. `sbt publishLocal` to build compiler jars and "publish" it into local repo where translator project could find it
3. Go to translator dir
4. `sbt run` (or load that project into IDE and run it there) to try autogenerating all possible specs:
  * Sources for KST specs are in `spec/kst`
  * Results would be in `spec/kst/out/$LANG`
  * One's expected to check generated specs and move it manually to `spec/$LANG` after checking that it actually does the job

**Comment by CWood1** [comment]
So, I've got the translator outputting tests now. Few compilation issues here and there due to incorrect syntax, nothing too extreme.

Next question: how exactly does one run the compiler against the test suite, then the tests against the compiled specs? Once I get these compiler errors resolved, that's my next goal is to have 50something failing tests I can then make pass.

**Comment by GreyCat** [comment]
@CWood1 Could you show me what exactly have you achieved so far?

**Comment by CWood1** [comment]
Okay, where we're at (I'll be pushing all this shortly, soon as I've got a couple compiler errors ironed out), I've got a rust library project under spec/rust. Inside of there, test/ houses all of the tests.

Therefore, open questions:
 - How do I get the compiler to run, and output the results into `spec/rust/src`?
 - Can I script updating `spec/rust/src/lib.rs`, to publicly declare all of the compiled modules in a way the tests can access?
 - Is there a framework by which I can automatically run the above, and then invoke `cargo test`?

Doing the above manually is relatively easy, though tedious. Are there ways by which I can script them, under existing frameworks? Or would that need to be developed from scratch?

**Comment by GreyCat** [comment]
> How do I get the compiler to run, and output the results into `spec/rust/src`?

There is a huge mess with terminology we have here ;)

If by "compiler", you mean ksc (i.e. app that translates ksy files into parsing source code in target language), then normally, it is run with [build-formats](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/build-formats) script, it runs the compiler on all formats for all target languages, and outputs results into `compiled/$LANG`.

If by "compiler", you mean kst (i.e. app that translates kst files into test specs in target languages), then it is right now runnable by invoking `sbt run` in its dir (or, equivalently, launching its main class in IDE).

I suspect there's slight misunderstanding on what needs to be done. Actually we have 3 "places" to update:

* Create a `RustTranslator` in [io/kaitai/struct/translators in compiler project](https://github.com/kaitai-io/kaitai_struct_compiler/tree/master/shared/src/main/scala/io/kaitai/struct/translators); that would be "expression translator", which translates KS expression into Rust expression (typically expression is translated into a single string, but recently we've got support for languages like Go which require translation of expressions into multiple statements).
* Create a `RustCompiler` to [io/kaitai/struct/languages in compiler project](https://github.com/kaitai-io/kaitai_struct_compiler/tree/master/shared/src/main/scala/io/kaitai/struct/languages) and register it [here](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/languages/components/LanguageCompilerStatic.scala#L12-L25)  — this would be needed to actually run `ksc -t rust` and be able to output any target language code from a given .ksy file, i.e. perform the normal function of a compiler.
* Add a `RustSG` spec generator to [io/kaitai/struct/testtranslator/specgenerators in KST project](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/translator/src/main/scala/io/kaitai/struct/testtranslator/specgenerators). This would rely on `RustTranslator` to generate the specs.

That's why I asked you to show what you've got, so I could understand which of these 3 things you were working on :)

> Can I script updating `spec/rust/src/lib.rs`, to publicly declare all of the compiled modules in a way the tests can access?

No equivalent of these exists so far in any other languages. C++ might benefit from similar automation, but so far we've just [added all tests manually to CMake project](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/cpp_stl/CMakeLists.txt). This is actually a somewhat sloppy moment, but for compiled languages you don't usually have an option to go with "build everything that would build, and run the tests with everything's that we've built". Normally (for C#, C++, Java, Go), a single broken file ruins the build, so we employ various tricks to reduce that risk. Having a manually maintained list of tests that "should run ok" is a first step to mitigate that problem.

> Is there a framework by which I can automatically run the above, and then invoke `cargo test`?

[ci-all](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/ci-all) script invokes all `ci-*` scripts in sequence. We should probably create `ci-rust`, which would run thte tests and deliver some kind of (ideally, JUnit-like XML) report to `test_out/`.

**Comment by CWood1** [comment]
As a progress update on the above, I used Cargo's build script facilities to automatically copy over everything from `compiled/rust` into `spec/rust/src`.

It first clears the `src` directory. Then, one by one it copies the files across, building `lib.rs` in the process. This removes the need to manually specify `lib.rs`, as it now updates itself every time the test suite is executed.

**Comment by GreyCat** [comment]
> As of right now, I'm yet to work out how to represent `nullAssert` in Rust,

Null assert is a comparison of some nullable values with null. "Nullable" are mostly the values that have `if` condition which can be false, and thus there would be no value, and typically we have some sort of `null` equivalent returned there.

> and am entirely unsure what `trueArrayAssert` is, however these will be worked out in time.

"True" arrays are arrays which result from `repeat`-style constructs, i.e. arrays of some type. Usually they are represented by some sort of `ArrayList`, `std::vector` or similar dynamically resizable generic array type.

Opposed to them, "non-true", byte arrays are a special type which usually results from just specifying `size: X` without any `type`. They are typically special, as it's possible to handle them in much more efficient manner: many languages allow some kind of `byte[]` non-dynamic array or `String`-like type with strings of bytes.

Typically, comparison and assertion of "byte arrays" is done in exactly the same manner as comparison of other primitive types, but comparison of arrays frequently requires some special call like "assertDeepEquals" or something like that, that will compare values inside the container, not just the identity of containers.

**Comment by GreyCat** [comment]
> It may even be worth attempting to update the test translator itself, to allow for outputting of multiple files. Under this framework, it would be possible to create an individual crate for every file under test. This eliminates the problem of a syntax error in one file entirely halting the testnig procedure, and other compiled langauges would also benefit from this as well.

Test translator already outputs one file per test/spec. The idea to compile every file as a separate test suite and launch it separately is definitely a solution, but its usefulness would vary from one language to another. For example, running ~140 executables for C++ is possible, but for Java it would be *very* slow. Right now, a test run takes ~1.4 s, but launching separate JVM for every suite would be 140x times slower => i.e. ~3 minutes.

We actually employ some crazy techniques like the one you can see in [run-java, lines 31-67](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/run-java#L31-L67), i.e. we try to compile everything in one go, but if it fails, we try compiling each file in separate compiler invocation, which is obviously very slow (i.e. ~140x times slower), but it allows us to isolate the culprit(s), mark it/them up as "failed to compile" and proceed with running the rest of the tests.

If you could think of better alternatives, especially cross-language, that would be most welcome. So far the only way I see is creation of some sort of "universal build" system, which could handle all languages, but this is clearly a huge overkill for a task like this.

**Comment by CWood1** [comment]
Okay, now we have a major issue. Because of the way Kaitai works, it expects structs/types/classes/whatever you want to call them, to be nested. Becuase of the way Rust works, they can't be. I'm yet to work out a solution to this.

Initially my thoughts were to collect each struct into some class somewhere, and have each class output itself in turn, but as far as fields go, there's no indication of what belongs to what. So I'm not entirely convinced.

**Comment by GreyCat** [comment]
@CWood1 That's not a really major issue, quite a few languages work that way. There are 2 distinct problems here, I'm not sure both apply to Rust:

* If you don't have any kind of "nested" class names in a language (for example, like in Go or Construct), you can just dump class names like `foo::bar::baz` always as full names like `Foo__Bar__Baz`. Every ClassSpec has a `name` property, which is guaranteed to have a proper full name of class with all nesting components. For example, [here is how it is used for Construct](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/ConstructClassCompiler.scala#L126)

* If you are not satisfied with normal call order which generates stuff like

```
class Foo {
  class Foo__Bar {
    class Foo__Bar__Baz {
    }
  }
}
```

you can set up `innerClasses` to false, for example, [as PerlCompiler does](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/languages/PerlCompiler.scala#L25). This would result in an order of calls to be like that:

```
class Foo {
}
class Foo__Bar {
}
class Foo__Bar__Baz {
}
```

This `innerClasses` thing actually affects the order of calls in [compileClass method](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/ClassCompiler.scala#L45-L97). If you want something completely different, you can always subclass ClassCompiler (for example, we do that for Go), and implement it differently.

**Comment by CWood1** [comment]
Is there a way I can get it to output fields before methods? Because of the way Rust works, it needs

```
struct Foo {
// fields
}

impl KaitaiStruct for Foo {
// methods
}
```

**Comment by GreyCat** [comment]
Yes, it's slightly more complicated. To achieve that, one needs to subclass ClassCompiler and override compileClass method with different order of calls.

Go actually [does exactly that](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/GoClassCompiler.scala#L17-L49) — i.e. "class declaration" [comes first](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/GoClassCompiler.scala#L31-L36) and a `Read()` method and instance methods [come after that](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/GoClassCompiler.scala#L38-L41).

In order to register a ClassCompiler-derived class, though, one needs to [add it to Main.compile](https://github.com/kaitai-io/kaitai_struct_compiler/blob/fb0e001e4c62a53316790a9e145c3bff0d1fe234/shared/src/main/scala/io/kaitai/struct/Main.scala#L60-L61) as well.

**Comment by CWood1** [comment]
For full names, such as `Foo__Bar`, how does one then use that as a type? In that example, the type used in the `Foo` struct definition is `Bar`, and I've not the foggiest how to attach `Foo__` to that.

As a general update, however, `RustClassCompiler` has now been defined, and compilation order works. A slew of errors have been ironed out, and progress is generally getting made on all fronts. In theory, the test suite should compile soon.

**Comment by GreyCat** [comment]
> In that example, the type used in the `Foo` struct definition is `Bar`,

Could you show me the exact code that generates `Bar` and not `Foo__Bar`?

> A slew of errors have been ironed out, and progress is generally getting made on all fronts.

That's really cool! Would you mind sharing your branch in public — may be I can review it and throw it some ideas, and may be help with integration into the CI?

**Comment by CWood1** [comment]
The relevant code is in [https://github.com/CWood1/kaitai_struct_tests](tests) and [https://github.com/CWood1/kaitai_struct_compiler](compiler). I'll post links to the name stuff tomorrow, when I'm not about to sleep :P

Thank you once again, so very much, for your help in this project. I couldn't have gotten nearly this far without you. You've been truly invaluable. Thank you.

**Comment by GreyCat** [comment]
@CWood1 What would you think about merging your current effort into master, and making it appear in [CI](http://kaitai.io/ci/) and tests?

> Clearly it must be possible to have a special comparison for things such as vectors, for languages like Java and C++, though I'm yet to find it. If someone can point me to that before I find it manually, I'd be grateful, but until then I'll keep looking.

There's no good magic there, this is how JavaCompiler does it:

* There is a stateful `var` named `switchIfs` — see https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/languages/JavaCompiler.scala#L480-L484
* When switch starts, `switchStart` is the first to be invoked, and it makes a decision which switch model to choose — i.e. either true switches or a series of ifs.
* If we chose "series of ifs", then if generates [it like that](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/languages/JavaCompiler.scala#L518-519), using helper [switchCmpExpr](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/languages/JavaCompiler.scala#L507-L514) method.

**Comment by CWood1** [comment]
> @CWood1 What would you think about merging your current effort into master, and making it appear in CI and tests?

So, at the minute, the tests don't acutally compile and run. I'd like to get the tests compiling first, if possible, because otherwise I fear I may actually break CI :stuck_out_tongue:. Feel free to disagree with me on that one, though.

I'll take a look into switchIfs shortly, and see how far I get with that. That should be enough to clear the next couple of hurdles at least.

Thanks so much for all of your help. Sorry things have slowed down considerably from the start of this project, I've started a new job which has my attention during the week, and I don't think they'd be too happy if I just carried on with my own stuff on the clock.

**Comment by CWood1** [comment]
I'm alive, just to let everyone know. Sorry about the long hiatus, personal things got in the way that took priority.

I've fixed the switch stuff, and that's now working. Among the errors I'm now getting are problems with `process_custom.ksy` - it tries to import `nested::deeply` and `my_custom_fx::MyCustomFx`. Naturally, these don't exist, which results in a compilation error. I checked the other tests, doesn't exist there either, yet somehow they pass, so clearly something is being done at some level to make that work.

Where should these exist? Are these just "magic" files that need hand writing and checking into the test repo? Or should they be generated somehow, somewhere?

**Comment by GreyCat** [comment]
@CWood1 I'm super glad to hear that you're ok and still motivated to continue ;)

`process_custom` is described [in the docs](http://doc.kaitai.io/user_guide.html#custom-process). This is indeed a way to invoke language-specific processing routine where it's needed, so these files you're missing are indeed hand-written. For example,

* For JavaScript: [MyCustomFx.js](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/helpers/javascript/MyCustomFx.js) and [nested-deeply/CustomFx.js](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/helpers/javascript/nested-deeply/CustomFx.js)
* For Java: [io.kaitai.struct.testformats.MyCustomFx](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/java/src/io/kaitai/struct/testformats/MyCustomFx.java) and [nested.deeply.CustomFx](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/java/src/nested/deeply/CustomFx.java).
* etc

We're actually developing something called [kaitai_compress](https://github.com/kaitai-io/kaitai_compress), which is a collection of wrapper routines related to data compression/decompression, implemented for different languages, compatible with this "custom processing" idea.

**Comment by GreyCat** [comment]
Sorry for delayed response.

> Roadmap so far:

I would actually think of getting your work merged into the master repo as top priority. Could you create a PR?

> Look into creating a kaitai-compress Rust crate

This is *very* experimental thing, so probably a very distant goal.

> On the subject of crates.io, how would one go about that? KaitaiStruct is not yet ready, and I've obviously not started the other one at all, but once they're formalised and v1.0, how would I get them on the repo under the Kaitai project?

https://github.com/kaitai-io/kaitai_struct_rust_runtime/ already exists under kaitai-io, I guess. I can add it into [main umbrella project](https://github.com/kaitai-io/kaitai_struct/) as a submodule whenever you'll say so.

> Does someone have to submit it under the project, and I don't worry about it? Or do I do it under my own email?

If you mean crates.io and similar submissions, we can create special account for that, I guess, something like crates@kaitai.io? That domain's mail is currently served by zohomail, so there would be a web interface for that and all that stuff...

**Comment by CWood1** [comment]
PRs has been created, and since my last message that's all the work I've had the time to do, I'm afraid.

That makes sense about the crate, on both points, I assume I can just ping someone on here or something and that'll get done? Or, failing that, I'd be happy to look after the Rust crate if needed, and run crates@kaitai.io accordingly. This is all a long way off either which way.

**Comment by berkus** [comment]
To publish just use the agreed email and run `cargo login / cargo package / cargo publish` - see [here](https://doc.rust-lang.org/cargo/reference/publishing.html).

**Comment by jonstewart** [comment]
Coming at this with no context, is there anything I could work on? I'd be very glad to have Rust support in Kaitai.

**Comment by CWood1** [comment]
Progress is being made, I'm sorry for the radio silence. I've been trying to fix the issue with the junit crate which has had me blocked. I'll be redoubling my effort in the new year.

**Comment by CWood1** [comment]
If I do a little bit of reorganising, I'm hoping to be able to get the initial version merged in sometime in January.

**Comment by jonstewart** [comment]
@CWood1 Thanks for the update! I'd be happy to help, especially if you can point me at starting places. I have a modicum of Scala experience, a smidge of Rust (with a desire to learn more), and a heavy background in C++ and dealing with binary data. I'm trying to get my team to adopt both Kaitai and Rust, so am eager to learn more and participate.

**Comment by CWood1** [comment]
So, to give full context to what's going on right now, cargo-test-junit appears to be dead. I've been trying other ways to get the xml file Travis needs, to no avail, so just now I've forked it. Reason being, one of its dependencies appears to be out of date. I'll get that updated and in, either today or tomorrow. At that point, updating the Kaitai repo to use the new dep will be nice and straight forward.

From there, process_custom hasn't been implemented at all, there's currently no way yet to walk back up the tree or process from root, and I'd like to look at getting the output `nostd` if I can, as a stretch goal. More work may end up surfacing as these are completed, but that's where we're at right now.

You probably won't be able to help very much until the tests are in and successfully working, though if my new way works, it won't be January, it'll be like tomorrow. Following on from that though, given you're not so familiar with Rust yet, process_custom will probably be easier, and I'll play with some weak pointers to try and get root/etc working.

Hang fire for now, I'm hoping I can speed this along and actually get some code merged (I'm getting impatient on that myself). Once that's done, things will pick up a lot more steam again, and hopefully not take too much longer to get this project finished.

**Comment by berkus** [comment]
Cargo.toml?

**Comment by GreyCat** [comment]
@CWood1 Basically, CI issues boil down to:

* We need have a way in [tests](https://github.com/kaitai-io/kaitai_struct_tests) repo to run all Rust tests.
  * It should produce `test_out/rust/ci.json` (and, optionally, some extra files in `test_out/rust`).
  * It should produce meaningful `ci.json` even when some of the files fail to compile and/or test errors occur
* We need to set up relevant martix steps in [ci_targets](https://github.com/kaitai-io/ci_targets) repo, I'll do  trivial stuff right now.

**Comment by CWood1** [comment]
Okay, that's about what I'd expect for the time being. Updated roadmap for testing:

- I've created cargo-junit, a fork of cargo-test-junit, to fix some bugs as the other one was abandoned.
- I need to unit test cargo-junit.
- I need to separate out compilation errors into separate errors (not sure if I should have one error per file, one error per compilation error, or an option to switch between the two).
- I need to find a way to `cargo install cargo-junit` as part of the test harness, before running the tests (to answer @berkus 's point, Cargo.toml is for dependencies in the source, not for altering the way it's built. cargo-junit extends cargo, with the intention of producing a junit style .xml output file).
- I'm toying with the idea of trying to run tests that compile, even when others in the suite don't. Right now it doesn't do this, so I'd need to come up with a way to enable this. I have some ideas, but I'm not sure how much of a priority this is yet.

**Comment by CWood1** [comment]
More updates:

- In my attempts to get `cargo-junit` into the state I need it to be in, in order for Kaitai to rely on it, I unearthed more issues. `test-to-vec` was made by the same engineer, and is in a similar state of abandonment, so I've forked it to `cargo-results`.
- `cargo-results` needs Suite to be cleaned up, so that it doesn't need so many (or ideally any) lifetimes.
- It should also be refactored.
- It should separate out compiler errors into separate error entries, by taking a flag - either, one error per failed file, or one error per error.
- `cargo-junit` should be refactored, which I'm in the middle of doing. This is in aid of the formalisation/unit test work, which is needed for a robust test suite.
- `cargo-junit` should take a command line arg to switch compile error modes in the output.
- `cargo-junit` should take a flag, to use `cargo test --all`, so it can work with multiple recursive Cargo directories.
- In the Kaitai test suite, infrastructure shall be written to give every test its own Cargo project, nested under a global one. This will then facilitate the final point above, in allowing for failing compilation while still testing what it can.

The joys of formalisation, the stack of work is growing as I find more issues. Once I hit this point though, we should have a pretty robust testing infrastructure.

**Comment by CWood1** [comment]
So, the testing work should now be in a usable state. There's a crate, `cargo-junit`, which is currently at version 0.8.0. This allows the output of a JUnit XML file. I've not done all the work I want to do on it, but it's enough for the purposes of this project now and enough that I'm not going to pursue it further for a bit because I need to get back to this.

@GreyCat would you be able to assist again in getting CI set up?

* `cargo test --all` is not implemented, and for now it won't be.
* As a result, the Kaitai test suite does not currently have a project per test, and for now it probably won't.

For now though, this will be enough to get us off the ground and having useful CI. I feel like this project is stalled pretty horribly, which I want to address, so I'm going to leave off testing for now and crack on with useful things again.

I've updated the CI runner in the `kaitai_struct_tests` repo, which I'll send a PR for now, to make it use `cargo-junit`, but I'm not entirely sure where to put the `cargo install` command. Once I've done that, it'll be back to implementing more Rust output features, and making actual headway again for the first time in what seems like forever.

I'm really sorry there have been so many breaks in my work on this. Graduation got in the way, then a new job, then moving flat, etc. Life has a habit of not wanting me to work on the things I enjoy. I'm hoping that'll be that for a good while now though.

**Comment by GreyCat** [comment]
@CWood1 Nice to hear from you again, welcome back :)

Regarding CI — I'll take a look and see if we can get your implementation running in a CI.

Please note that recently a new concept of "unified building" framework for statically compiled languages was brought in, and I'm slowly moving all compiled languages (C++, C#, Java, hopefully Go will be there soon) to it.

The concept is very simple: actually, building tests for all these languages have something in common. We generally implement the same loop for all of them:

1. Try to build
2. If it fails, try to remove some files from the project and retry
3. If we're unable 

You can read more at [tests/builder](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/builder). If we'll go this way with Rust, typically it means that one needs to implement a class with [these 7 methods](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/builder/partial_builder.rb#L186-L234).

Also note that there is some incoming work from @bspeice: https://github.com/kaitai-io/kaitai_struct_compiler/pull/161#issuecomment-484877046 — could you guys discuss whatever needs to be done & coordinate the efforts? I'm kind of hesitating merging anything into compiler as doing this one-sidedly will ruin other person's work.

**Comment by bspeice** [comment]
Apologies that I hadn't seen the history here; I'm mostly interested in having something that works, so I'll be thrilled to kill off my code if there's a good solution elsewhere.

My biggest concern is that the `Option<Box<Struct>>` types don't make sense for `_parent` and `_root`. You can't have multiple people owning `_root`, but everyone needs access to it. If you wanted to avoid reference lifetimes, `Option<Rc<Struct>>` is workable, but I'd rather just use references to avoid needing heap allocations. There's a couple other issues I mention in kaitai-io/kaitai_struct_compiler#161, but I don't view those as anything more than implementation details.

**Comment by GreyCat** [comment]
@CWood1 
> but I'm not entirely sure where to put the `cargo install` command

Typically it goes into `prepare-$LANG` file in ci_targets. Here's [prepare-rust](https://github.com/kaitai-io/ci_targets/blob/master/prepare-rust) we currently attempt to run in CI.

**Comment by LogicAndTrick** [comment]
In regards to the `Option<Box<Struct>>` stuff, those types were pure guesses based on my fumbling through the Rust book. I never got parent/root references working, so the types are just placeholders really. Please don't hesitate in changing them!

**Comment by CWood1** [comment]
@bspeice - I'm not sure how far along you are in the various feature sets, so I'm hesitant to suggest going with either of our projects yet. Lets compare notes, it might make sense to merge features in from both.

As for `_parent` and `_root`, that's a bug in my personal backlog, that I'm yet to get to. I'm not yet sure on the solution, but it's one that's starting to near the top.

@GreyCat - Noted, I'll take a look at that. I'm sort of going to have to leave it here for today, but I'll pick that up tomorrow, as well as the build artifacts I accidentally committed that you pointed out in the review (again).

**Comment by bspeice** [comment]
@CWood1 - Sounds like a plan. I'll clean up what I can in the implementation I had, and then I think we can work from there!

**Comment by CWood1** [comment]
That sounds like an idea to me. Let's use this issue to track the complete
project, so it's all in one place and we can collaborate a little better

**Comment by CWood1** [comment]
Right now, I'm trying to figure out how to get `parent` and `root` to work. Rust doesn't like circular references, so I'm not sure this is possible without an `Rc`. @bspeice I'd love to hear your suggestions.

Also, @bspeice, to save on the possibly duplicated work, what do you think of merging your work in with my repo, and co-authoring a PR? That way we're not coding on top of one another, potentially ruining each other's work as we work in parallel. Both compiler, test, and Rust runtime repos are fully up to date with what's on my hard disk, so we'll be working from a common point.

**Comment by GreyCat** [comment]
@CWood1 I see that you've updated your PRs. Would it be ok to merge them now?

@CWood1 @bspeice For the "unified" PR: guys, if that would be of any help, I'd be glad to invite you to the team and give you access to commit directly in compiler/tests/almost any other kaitai-io org repo. You can thus collaborate on a certain branch there, and it will serve as the main focus point?

**Comment by CWood1** [comment]
Re updated PRs, not yet. I'm currently trying to figure out how to get
`parent` and `root` to work without making the borrow checker cry.

Re becoming team members and having a dev branch, I'm up for that approach
as well. No real preference to the approach, I was mainly trying to
establish a common point rather than us both having our own PRs and coding
on top of one another.

**Comment by GreyCat** [comment]
Heh, that `_parent` / `_root` circular deps were something that @LogicAndTrick I believe hit the roadblock with originally.

@CWood1 Please also note this: https://github.com/kaitai-io/kaitai_struct_tests/pull/52/files#r276985355

**Comment by CWood1** [comment]
For the tests - aye, still need to clean that up. On my radar, I just
didn't have chance to today. I'm slightly on holiday this week,I just had
an idea so quickly logged in so I could bring everything into line for
collaborative purposes. I'll do that probably Saturday.

**Comment by bspeice** [comment]
In this case, `Rc` isn't necessary. Storing `_parent: Option<&'a Self::Parent>` works fine because the borrow checker only needs to validate that `_parent` lives at least as long as the children; the way other libraries like `nom` typically do this is to tie the lifetime of everything being parsed to the lifetime of the stream. Then, it's just a matter of convincing the checker this we actually did it correctly (which admittedly wasn't fun when I worked through it earlier).

However, this does lead to a slightly different API; Kaitai's C++ runtime currently heap-allocates vectors for everything, where `nom` and other parsers in Rust typically take a zero-copy approach and return byte slices for everything. See my PR (kaitai-io/kaitai_struct_compiler#161) for more detail.

My preference (for the moment) is to stick with a separate PR because I'm going a slightly different route for parsing that makes it difficult to integrate with the existing code (as well as requiring a different runtime library). Haven't taken a look at testing though, so I should definitely do that soon.

**Comment by CWood1** [comment]
@bspeice - suggest at some point we sit down and compare notes thoroughly.
I'd like to understand your approach in depth, and explain mine similarly,
so we can avoid unnecessary duplicate effort. Would you be able to put
something like that together please? I'll do the same Saturday. Doesn't
need to be thorough, I'm just keen to merge this into one work stream so we
can collaborate rather than overwrite one another's code and create more
work for each other.

**Comment by bspeice** [comment]
@CWood1 - Totally understood. I'm working towards making sure the approach is viable in the first place, and then happy to write up what's going on. It's a whole lot easier to do with examples.

**Comment by GreyCat** [comment]
Just to reiterate: if I can help you guys with anything — examples, tests, infrastructure, explaining any architectural choices we've made so far, etc — please don't hesitate to ping me.

**Comment by bspeice** [comment]
Architecture question: current implementations don't seem to handle "we ran out of data" well. Is it just up to the user to reset the stream position in these scenarios?

**Comment by CWood1** [comment]
Fwiw the most elegant solution that immediately comes to mind would be to
reset the stream position ourselves, then error with a predefined
"insufficient data" error. There are cases - network streams for instance -
in which the right response is wait for data and try again, and other
instances - files, for example - where an error should be returned from the
tool using whatever grammar. Wdyt?

I've seen your PR, and while I haven't had time to look through it in
depth, I like the way you handle lifetimes at first glance. I'm not
entirely sure how you handle streams yet, that bit isn't clear to me
without closer inspection which I haven't been able to do.

I'm not entirely sure I agree with having the stream of the parsed struct
be tied to the lifetime of the stream though. Not unless we make `_root`
the owner of the stream. Otherwise the user will be required to keep dead
streams around for no reason. Say, loading in a file, parsing, but then
needing to keep it around for the lifetime of the entire tool running.
Given kaitai is run once to parse and there's your structure, I'm not sure
I agree with tying the lifetime of the struct to the lifetime of the
stream. Wdyt?

**Comment by bspeice** [comment]
On stream recovery: had a quick conversation with @GreyCat on Gitter, and that was where I came to; it's probably up to the user to save the stream position (using `_io.pos` or similar) before parsing, and then seeking back to that if desired. Think we're both agreed.

And I spoke incorrectly earlier on lifetimes, the borrow checker thoroughly convinced me of the error of my ways. What I should have said was the lifetime of the struct must be bounded by the lifetime of the stream (that is, the stream must live at least as long as the struct). I'm against `_root` owning the stream, as very often the same stream will be used to parse multiple root structures (network streams).

**Comment by CWood1** [comment]
Point taken on one stream being used for multiple structs. I'm still not
entirely sold on having the struct lifetime bounded by the stream lifetime
though. If, say, a struct gets returned across a library interface, we now
need another return value, which is just "you don't need to touch this,
just save it somewhere", which seems a touch inelegant and abstraction
leaky. It may not be possible to avoid it, I'm not sure without digging
into your design more, but I'd rather avoid it if possible for elegance and
better abstractions, if that makes sense. Thoughts?

**Comment by bspeice** [comment]
Yeah, returning across lib boundaries is going to be painful because the structs I've been creating don't own their data (unlike other Kaitai implementations). And now that I think about it, `#[derive(Clone)]` won't work because they'd just copy the `_parent` and `_root` references.

My personal use case is interested in having the zero-copy style, but maybe we make that a compile option to choose between owning and non-owning codegen?

**Comment by CWood1** [comment]
A compile option seems like quite an elegant way to do this actually. I
take your point about zero copy too, so tradeoffs. But, if we can support
both, why not. Would that be baked into kaitai itself do you think, or as a
feature flag in the rust output?

On a related note, I have a vested interest in making the output support
nostd, as one of my plans for this would be in my kernel. Happy to talk
about this in more depth at a later date on im, and we can go over design,
but this will probably influence the design a little. For the most part it
will mostly be a feature switched reimport of std into core. Think nom does
the same?

**Comment by bspeice** [comment]
On compile options, baked into the Kaitai compiler, it basically switches structs from using `&[u8]` in the non-owning case to `Vec<u8>` in the owning case.

For `nom`'s `no_std` support - [sure looks like it](https://github.com/Geal/nom/blob/5764a02d20bed7a8b44d0d014eb20806f8589117/src/lib.rs#L372-L399). `no_std` would be difficult (both `process_*` functions in the runtime and conditional repeats in `read` need heap allocation), but we can feature-gate those and let users deal with the compilation issues.

Finally, having learned my lesson on the borrow checker coming out with some nasty surprises, I've got a pretty basic test case available [here](https://github.com/bspeice/kaitai_rust/blob/master/tests/lifetime_validation.rs) that doesn't yet compile; if we can get this working (currently has issues with proving validity of `_parent` and `_root`) then I think we're in the clear. It may require some API changes (giving `_root` and `_parent` references to `read` instead of storing in the struct seems like a good idea), but still making progress.

**Comment by CWood1** [comment]
Re `no_std`, heap access is still possible with `alloc` crate. It's just a
little more limited than full `std`, but I honestly don't think it'll be a
problem. Might be necessary to have a slightly limited feature set to
support `no_std`, but that'll be it.

Re your test case, I'll pitch in again on Saturday if it still doesn't
compile. I can't get to my workstation until then I'm afraid, but I'm happy
to pitch in from there. I'm getting more and more sold on your approach to
solving the circular reference issue, I quite like the reference approach.
If there's a way I can sync my work into that, I will.

Would it be worth considering as well initialising member structs from the
parent read, rather than pushing that responsibility onto the caller?
Otherwise it's much more boilerplate, which I'm keen to avoid from an API
design standpoint. In my ideal world, it should be possible to say
Struct::read(std::io::Cursor), or similar, and that initialises the whole
Struct.

**Comment by bspeice** [comment]
The more I work with it, the more I'm convinced we *have to* initialize members in the parent read. Because self-referencing is such an issue (to the point where I'm not sure it's possible) I think the current class needs to own its children, and the `read()` signature becomes something along the lines of `read(stream: &S, parent: Option<&P>, root: Option<&R>) -> Result<Self>`, which looks strangely like how `nom` handles things (:wink:). That said, adding a `read(stream: &S)` that sets things up for the root-level structs and then calls each child `read` would likely be pretty easy to implement on top.

Should have some interesting stuff drawn up tomorrow.

**Comment by bspeice** [comment]
Alright, new issues. Because `instance` definitions can use `_parent` and `_root`, and we compute instances lazily, we've got conflicts with lifetimes. It's not a pretty API, but my plan for now is to always require `_root` and `_parent` being passed into the instance retrieval method, but we'll need to figure out a better API for the future.

**Comment by CWood1** [comment]
Would you be able to give an example of that? I'm not quite sure what the
conflict is right now.

**Comment by bspeice** [comment]
The [example](https://github.com/bspeice/kaitai_rust/blob/master/tests/lifetime_validation.rs) I posted yesterday is good enough to demonstrate the issue. There's two things that are problematic:

1. In the version where parents only store references to children (ex. websocket message storing a vector of its dataframes), we can't create new child structs during `read()` because they'll be dropped at the end. Fairly straight-forward, nobody owns the children so they disappear.

2. If you modify it such that parents do own their children, there are a ton of issues that show up during `read()` that I admittedly don't quite understand.

The example `read` method:
```rust
    fn read<'s: 'a, S: KStream>(&mut self, stream: &'s mut S) -> Result<(), KError<'s>> {
        self.bytes = stream.read_bytes(1)?;
        let mut child = TestChildStruct::new(Some(self), Some(self.root()));
        child.read(stream)?;
        self.child = Some(&child);
        Ok(())
    }
```

The first issue has to do with `stream` being delivered to the child, which I admit I'm a bit surprised by, since I would've expected this is legal:
```
error[E0499]: cannot borrow `*stream` as mutable more than once at a time
  --> tests/lifetime_validation.rs:38:20
   |
34 |     fn read<'s: 'a, S: KStream>(&'a mut self, stream: &'s mut S) -> Result<(), KError<'s>> {
   |             -- lifetime `'s` defined here
35 |         self.bytes = stream.read_bytes(1)?;
   |                      ---------------------
   |                      |
   |                      first mutable borrow occurs here
   |                      returning this value requires that `*stream` is borrowed for `'s`
...
38 |         child.read(stream)?;
   |                    ^^^^^^ second mutable borrow occurs here
```

The second definitely makes sense as being illegal. When we create a new child struct, we borrow `self` and give it to the child (which also effectively borrows `self.child`). We then assign `self.child` later with the owned value which causes issues, because that assignment needs to borrow `self` as well.

```rust
error[E0506]: cannot assign to `self.child` because it is borrowed
  --> tests/lifetime_validation.rs:39:9
   |
20 | impl<'a> KStruct<'a> for TestRootStruct<'a> {
   |      -- lifetime `'a` defined here
...
37 |         let mut child = TestChildStruct::new(Some(self), self.root);
   |                         -------------------------------------------
   |                         |                         |
   |                         |                         borrow of `self.child` occurs here
   |                         argument requires that `*self` is borrowed for `'a`
38 |         child.read(stream)?;
39 |         self.child = Some(child);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `self.child` occurs here
```

There's a total of seven errors that show up in the mini example I gave. I'm simultaneously glad because that example is catching a lot of issues early, but they're still problematic.

**Comment by bspeice** [comment]
OK, lots of updates and some good news. I have a definition that passes the borrow checker, works for all the simple structures I can think of, and lets us scale into the future. I'm documenting how I went through the process in case anyone else is crazy enough to try re-building this with a different structure in the future.

First, a [super-simple sketch](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4276b7a06a7a785f6a99c846464c10b2) of what we're ultimately after. We need to read things in, have nested structures (and thus deal with ownership) while maintain zero-copy access to the underlying stream.

From there, we [try to add](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=096e9789cd56442e8571f7eae78629bd) a `KStream` trait, and pass that as mutable to the readers. That doesn't go so well because we can't give the child struct access to the stream while we maintain a mutable borrow ourselves, and fancy scoping doesn't help at all.

So, we need to make the streams immutable. Thankfully, Rust has a trick for dealing with this - the `RefCell` type. If we [move the stream state](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=212e3c454033c1fb449430516128f4de) into the cell, we can stay no-std compliant (and no heap allocation needed). We'll have to do a lot of `borrow` and `borrow_mut` calls in the stream definition, but I'd rather pay that cost than have structs own their data and have to allocate everywhere (no actual benchmarking was done to support this conclusion, I'm just imitating other well-known libraries).

Finally, we add in the [definition](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bfcd1035cf3bba8e236447679502f173) of `KStruct` so that all structs have a common language for instantiating/reading each other.

The result is the API available here: [kaitai_rust](https://github.com/bspeice/kaitai_rust/blob/01d42890f6b4c14a2f7e927220edf09041bd8ba9/src/lib.rs). It includes a couple things not available in the Playground (a definition of `KError` for example), and a [fixed version](https://github.com/bspeice/kaitai_rust/blob/01d42890f6b4c14a2f7e927220edf09041bd8ba9/tests/lifetime_validation.rs) of the example I posted previously.

Now, there are a couple of issues that need to be addressed before it's considered `done`.

1. How do we handle `instance` definitions? I need to actually prove it out, but as long as we provide `_root` and `_parent` references the same way as in our `read` definition, I don't expect we'll run into issues.

2. How do we handle `process` directives? I'd like the structs to keep an owned buffer that can be provided, rather than how existing implementations allocate new buffers. This is a codegen issue though, not lifetimes, so not terribly worrisome.

3. Support for `_root.io` directives in instance definitions? I'm not totally familiar with what's going on here, so not sure how hard it'd be, but it's also not super-prevalent so I'm not treating as a big concern at the moment.

Those concerns aside, it's not unreasonable to think that there's a Rust implementation on the horizon.

3.

**Comment by bspeice** [comment]
Big news! A simple Websockets-based example is now running, so at the very least, a PoC compiler is ready. Following up from the last post, `instance` definitions are indeed handled by passing in `_root` and `_parent`, I need to see if there's a better API we can put on top of that, but they're functional. `process` directives are still up in the air. And `_root.io` shouldn't be too bad - from what I can tell, it's just some extra function calls around `pos()`, `seek()`, etc. so I don't think there will be too many worries.

Now, it's time for a couple last cleanup things. First, the code isn't something I'm super proud of, so a refactoring pass is in order. Second, we don't have a good way of handling calculated endianness, similar to Go. Finally, I have no idea what else will break as soon as we start trying other types, the Websocket message is a reasonably simple format.

Still, progress is progress.

**Comment by GreyCat** [comment]
@CWood1 @bspeice Can I ask you guys to still devote some time to get us visibility of how well we fare in tests counts in CI?

Right now I have a feeling like a big chunk of tremendous work done by @bspeice is going virtually unnoticed due to (1) no consensus on merging, (2) no visibility in CI.

**Comment by bspeice** [comment]
Definitely. Now that I have a design that I know will work well, I think getting CI on it makes a ton of sense. For merge consensus, I'll delegate to you guys - is it worth looking to integrate the compiler/runtime I've been working on and then fix things in CI as we go? Or would we rather have testing done beforehand?

**Comment by bspeice** [comment]
@CWood1 / @GreyCat - Can you guys run me through how to generate the tests? I need to make some changes to the translator as part of the new API, but I'm unable to figure out how to run it locally.

**Comment by bspeice** [comment]
@GreyCat followed up via chat; `sbt publishLocal` for the compiler, then `sbt run` for the translator.

**Comment by sealmove** [comment]
> @CWood1 / @GreyCat - Can you guys run me through how to generate the tests? I need to make some changes to the translator as part of the new API, but I'm unable to figure out how to run it locally.

https://github.com/kaitai-io/kaitai_struct/issues/626

**Comment by ShadowJonathan** [comment]
Why hasn't this been implemented yet? Seeing rust absent from the gallery of other languages makes me not exactly inclined to use this for anything.

**Comment by rljacobson** [comment]
> Why hasn't this been implemented yet?

It looks to me like [it has](https://github.com/kaitai-io/kaitai_struct_compiler/tree/master/shared/src/main/scala/io/kaitai/struct/translators). Maybe @CWood1 or @GreyCat can tell us the status. When I tried to run it I got the following:

```bash
$ ksc -t rust png.ksy
png: /:
	error: AnyType (of class io.kaitai.struct.datatype.DataType$AnyType$)
```

**Comment by orangecms** [comment]
In case anyone seeks to improve the Rust support or its documentation, funding is available:
https://nlnet.nl/project/Kaitai-Rust/

**Comment by jonstewart** [comment]
That's interesting. I know a contract developer who's done significant work
on this. How much funding is available and what's the process? Or, maybe
better, who should he contact to discuss further?

**Comment by generalmimon** [comment]
The most recent efforts to add Rust were from @Agile86 and @revitalyr in these pull requests:

- https://github.com/kaitai-io/kaitai_struct_compiler/pull/250
- https://github.com/kaitai-io/kaitai_struct_tests/pull/96
- and also in this repo: https://github.com/Agile86/kaitai_rust/tree/rust_basic_support_v2

The [grant from NLnet](https://nlnet.nl/project/Kaitai-Rust/) has multiple parts, one of them is to follow up on these efforts with the goal of integrating them to Kaitai Struct. This includes adding missing pieces of test infrastructure (partial builder), reviewing the code, aligning it with KS standards and conventions, fixing issues etc.
